toolchain/parser/parser.cpp

```diff
@@ -733,25 +745,40 @@ auto Parser::HandleDeclarationLoopState() -> void {
       break;
     }
     case TokenKind::Interface(): {
-      PushState(ParserState::InterfaceIntroducer());
+      PushState(ParserState::APIIntroducer());
```

Comment:
`APIIntrodcuer` is my attempt to choose a name generic enough to capture interfaces and classes. Let me know if you have better names :).

Reply:
First, I'd suggest something more like `TypDeclaration` when talking about things like this. It's declaring a type, and both "API" and "introducer" are ambiguous terms ("API" should include things like "fn", "introducer" can also refer to things like "virtual").

Note though, the syntax of the two diverge. Comparing in parser.ypp:

https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/syntax/parser.ypp#L1165
```
| class_declaration_extensibility CLASS identifier type_params class_declaration_extends LEFT_CURLY_BRACE class_body RIGHT_CURLY_BRACE
```

https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/syntax/parser.ypp#L1194
```
| INTERFACE identifier type_params LEFT_CURLY_BRACE interface_body RIGHT_CURLY_BRACE
```

In this model, how do you plan to handle keywords before `class` as well as `extends`?

I'll also note, I'd rather you not write this to be so dependent on stack_context_ instead of using ParserState. Where you use stack_context_ presently in function definition handling, that made some sense to me because there's a fair amount of indirection and we'd need to rewrite a bit in order to get rid of stack_context_. However, stack_context_ is really going to need to be another stack of parser state in addition to the parser state stack -- and in that context, I'm hesitant to rely on it where we don't need to.

Since there's no indirection needed here (that is, you're directly controlling the behavior of APIIntroducer) I think it'd be better to use different states instead of relying on stack_context_. That also keeps the door open to refactor and remove stack_context_ in favor of distinct states, at lower cost.

That also means that as behavior diverges (e.g., `extends`) it's something specific to the `class` state, not necessarily in the common handler. As we do have more type declarations (mixins, choices, constraints) my leaning is to have a helper method for common logic, rather than a single state trying to handle all of them.

Reply:
Thanks for the extensive comment. I agree with your observations. I will abandon this PR and start a new one with more separate handling for classes and interfaces instead.